src/main/java/com/shapesecurity/shift/es2017/parser/PatternAcceptor.java

```diff
@@ -0,0 +1,728 @@
+/**
+ * Copyright 2018 Shape Security, Inc. <p> Licensed under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License. You may obtain a
+ * copy of the License at <p> http://www.apache.org/licenses/LICENSE-2.0 <p> Unless required by
+ * applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See
+ * the License for the specific language governing permissions and limitations under the License.
+ */
+package com.shapesecurity.shift.es2017.parser;
+
+import com.shapesecurity.functional.F;
+import com.shapesecurity.functional.data.ImmutableSet;
+import com.shapesecurity.functional.data.Maybe;
+
+import javax.annotation.Nonnull;
+
+import java.util.*;
+
+public class PatternAcceptor {
+
+    public final String pattern;
+    public final boolean unicode;
+
+    private static final String[] decimalDigits = new String[]{"0", "1", "2", "3", "4", "5", "6", "7", "8", "9"};
+    private static final String[] octalDigits = new String[]{"0", "1", "2", "3", "4", "5", "6", "7"};
+    private static final String[] hexDigits = new String[]{"0", "1", "2", "3", "4", "5", "6", "7", "8", "9", "a", "b", "c", "d", "e", "f", "A", "B", "C", "D", "E", "F"};
+    private static final String syntaxCharacters = "^$\\.*+?()[]{}|";
+    private static final String[] syntaxCharacterArray = "^$\\.*+?()[]{}|".split("");
+    private static final String extendedSyntaxCharacters = "^$.*+?()[|";
+    private static final String[] controlCharacters = new String[]{"a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k", "l", "m", "n", "o", "p", "q", "r", "s", "t", "u", "v", "w", "x", "y", "z", "A", "B", "C", "D", "E", "F", "G", "H", "I", "J", "K", "L", "M", "N", "O", "P", "Q", "R", "S", "T", "U", "V", "W", "X", "Y", "Z"};
+
+    private static HashMap<String, Integer> constructControlEscapeCharacterValues() {
+        HashMap<String, Integer> map = new HashMap<>();
+        map.put("f", (int) '\f');
+        map.put("n", (int) '\n');
+        map.put("r", (int) '\r');
+        map.put("t", (int) '\t');
+        map.put("v", 0x11); // \v in javascript
+        return map;
+    }
+
+    private static final HashMap<String, Integer> controlEscapeCharacterValues = constructControlEscapeCharacterValues();
+
+    private static final String[] controlEscapeCharacters = controlEscapeCharacterValues.keySet().toArray(new String[0]);
+    
+    private class Context {
+        private int index;
+        private ImmutableSet<String> backreferenceNames;
+        private ImmutableSet<String> groupingNames;
+        private ImmutableSet<Integer> backreferences;
```

Comment:
We don't need to keep all of these in a set, we just need to keep the maximum value we've seen.